manager/encryption/encryption.go

```diff
@@ -0,0 +1,130 @@
+package encryption
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+	"strings"
+
+	"github.com/docker/swarmkit/api"
+	"github.com/gogo/protobuf/proto"
+)
+
+// This package defines the interfaces and encryption package
+
+const humanReadablePrefix = "SWARMMKEY-v1-"
+
+// ErrCannotDecode is the type of error returned when some data cannot be decoded as plaintext
+type ErrCannotDecode struct {
+	msg string
+}
+
+func (e ErrCannotDecode) Error() string {
+	return e.msg
+}
+
+// A Decoder can decrypt an encrypted record
+type Decoder interface {
+	Decode(api.MaybeEncryptedRecord) ([]byte, error)
+}
+
+// A Encoder can encrypt some bytes into an encrypted record
+type Encoder interface {
+	Encode(data []byte) (*api.MaybeEncryptedRecord, error)
+}
+
+type noopCoder struct{}
+
+func (n noopCoder) Decode(e api.MaybeEncryptedRecord) ([]byte, error) {
+	if e.Algorithm != n.Algorithm() {
+		return nil, fmt.Errorf("not an unencrypted record")
+	}
+	return e.Data, nil
+}
+
+func (n noopCoder) Encode(data []byte) (*api.MaybeEncryptedRecord, error) {
+	return &api.MaybeEncryptedRecord{
+		Algorithm: n.Algorithm(),
+		Data:      data,
+	}, nil
+}
+
+func (n noopCoder) Algorithm() api.MaybeEncryptedRecord_Algorithm {
+	return api.MaybeEncryptedRecord_NotEncrypted
+}
+
+// NoopCoder is just a pass-through coder - it does not actually encode or
+// decode any data
+var NoopCoder = noopCoder{}
+
+// Decode turns a slice of bytes serialized as an MaybeEncryptedRecord into a slice of plaintext bytes
+func Decode(encoded []byte, decoder Decoder) ([]byte, error) {
+	if decoder == nil {
+		decoder = NoopCoder
+	}
+	r := api.MaybeEncryptedRecord{}
+	if err := proto.Unmarshal(encoded, &r); err != nil || r.Size() != len(encoded) {
```

Comment:
`r.Size()` calculates the size that would be needed to marshal this object. But I don't think protobuf makes any guarantees about only being able to serialize things one way. I'm not comfortable with this check because any change in the way the generated code serializes records could make it impossible to decode previously-saved records. Also, if we ever add a new field to `MaybeEncryptedRecord` it might cause problems. I think it's better to rely on the error returned by `Unmarshal`.